---
title: "R Notebook"
output: html_notebook
---


```{r}
# Libraries

library(dplyr)
library(Seurat)
library(readxl)

```


# Load Data

```{r}
filename_metaData <- 'C:/Users/Sylvia/Dropbox (Partners HealthCare)/Sylvia_Romanos/scRNASeq/Data/Dexa_meta.xlsx'
filename_sampleParam <- 'C:/Users/Sylvia/Dropbox (Partners HealthCare)/Sylvia_Romanos/scRNASeq/Data/sample_parameters.xlsx'
metaData <- read_excel(filename_metaData)
sampleParam <- read_excel(filename_sampleParam)

sample_name <- metaData$Sample[3]

folder <- 'C:/Users/Sylvia/Dropbox (Partners HealthCare)/Sylvia_Romanos/scRNASeq/Code/Output/'
filename <- paste("C:/Users/Sylvia/Dropbox (Partners HealthCare)/Sylvia_Romanos/scRNASeq/Data/",sample_name,"_raw_feature_bc_matrix.h5",sep = "")

filter <- TRUE

if (filter){
    folder <- paste0(folder,sample_name,'/Filtered/')
} else if (filter == FALSE){
    folder <- paste0(folder,sample_name,'/Unfiltered/')
}


```
# Load Metadata
```{r}
metaData <- read_excel(filename_metaData)
sampleParam <- read_excel(filename_sampleParam)
```

# Load Data
```{r}
  
data <- Read10X_h5(filename, use.names = TRUE, unique.features = TRUE)
data <- CreateSeuratObject(counts = data, project = "BM", min.cells = 3, min.features = 200)

nFeature_RNA_list <- list(sampleParam$RNA_features_min[sampleParam['Sample'] == sample_name]
                            ,sampleParam$RNA_features_max[sampleParam['Sample'] == sample_name])
percent_mt <- sampleParam$percent_mt_min[sampleParam['Sample'] == sample_name]
```

# QC
```{r}
data[["percent.mt"]] <- PercentageFeatureSet(data, pattern = "^MT-")
  
  if (filter == TRUE){
    # nFeature_RNA (Unique features) are genes/cell
    data <- subset(data, subset =  nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mt < 15) # For some reason this no longer works with variables
  }
  
# Visualize QC metrics as a violin plot
VlnPlot(data, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 4)
  #print(plot)

# Visualize Feature-Feature Relationships
plot1 <- FeatureScatter(data, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(data, feature1 = "nCount_RNA", feature2 = "nFeature_RNA") 
CombinePlots(plots = list(plot1, plot2))
  #print(plot)
```

# Get Variable Genes

```{r}
norm_val <- sampleParam$norm_val[sampleParam['Sample'] == sample_name]
nfeatures_val <- sampleParam$nfeatures_val[sampleParam['Sample'] == sample_name]

# Normalize
data <- NormalizeData(data, normalization.method = "LogNormalize", scale.factor = norm_val)
  
# Identification of highly variable features
data <- FindVariableFeatures(data, selection.method = "vst", nfeatures = nfeatures_val) 
  
##We noticed that among those genes are immunoglobulin genes, which should not be present in this dataset, as plasma cells have been filtered out through magnetic selection. 

# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(data), 10)
top10
  
```
#Score for Cell Cycle gene expression

```{r}
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes
if (filter){
  data <- CellCycleScoring(data, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)
}
```

# PCA

```{r}
# Scale
data <- ScaleData(data, features = rownames(data)) 
  
# PCA
data <- RunPCA(data, features = VariableFeatures(object = data)) # Change defualt value. Different per sample 
```

```{r}
print(data[["pca"]], dims = 1:5, nfeatures = 5)
  
VizDimLoadings(data, dims = 1:2, reduction = "pca")

DimPlot(data, reduction = "pca")

print(DimHeatmap(data, dims = 1:6, cells = 500, balanced = TRUE))

print(DimHeatmap(data, dims = 7:12, cells = 500, balanced = TRUE))
```

# visualize Dimemsions
```{r}
#(data <- JackStraw(data, num.replicate = 100))
#(data <- ScoreJackStraw(data, dims = 1:20))
  
#JackStrawPlot(data, dims = 1:15)
  
ElbowPlot(data)

```

# Cluster
```{r}
# Value of the resolution parameter, use a value above (below) 1.0 if you want to obtain a larger (smaller) number of communities.
resolution_val<- sampleParam$resolution_val[sampleParam['Sample'] == sample_name]
PCA_dim<- sampleParam$PCA_dim[sampleParam['Sample'] == sample_name]
data <- FindNeighbors(data, dims = 1:PCA_dim)
data <- FindClusters(data, resolution = resolution_val)
head(Idents(data), 5)
data <- RunUMAP(data, dims = 1:PCA_dim)

DimPlot(data, reduction = "umap")


```
#Visualize clustering
```{r}
print(FeaturePlot(data, features = c("S.Score", "G2M.Score", "nCount_RNA", "percent.mt")))

```
# Cluster Biomarkers
```{r}
# Find Cluster Biomarkers
  # find markers for every cluster compared to all remaining cells, report only the positive ones
invisible(markers <- FindAllMarkers(data, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25))
  #BM_filtered_markers  %>% group_by(cluster) %>% top_n(n = 2, wt = avg_logFC)
markers  %>% group_by(cluster) %>% top_n(n = 10, wt = avg_logFC)
  
  # Plotting the top 10 markers for each cluster.
top10 <- markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_logFC)
  
plot = DoHeatmap(data, features = top10$gene)
```

# Regress out Variables
```{r}
# Regress out percent MT and nCount_RNA because they appear to drive clustering
data <- ScaleData(data, features = rownames(data), vars.to.regress = c("nCount_RNA", "percent.mt"))

```

# Repeat Cluster
```{r}
## Rerun PCA
data <- RunPCA(data, features = VariableFeatures(object = data)) # Change defualt value. Different per sample

# Value of the resolution parameter, use a value above (below) 1.0 if you want to obtain a larger (smaller) number of communities.
resolution_val<- sampleParam$resolution_val[sampleParam['Sample'] == sample_name]
data <- FindNeighbors(data, dims = 1:PCA_dim)
data <- FindClusters(data, resolution = resolution_val)
head(Idents(data), 5)
data <- RunUMAP(data, dims = 1:PCA_dim)

DimPlot(data, reduction = "pca")

DimPlot(data, reduction = "umap")


FeaturePlot(data, features = c("S.Score", "G2M.Score", "nCount_RNA", "percent.mt"))

```

```{r}
# 
#   sample <- data
#   cell_list <- list(
#     'bcell_activated',
#     'bcell_plasma',
#     'bcell_memory',
#     'bcell_marginal',
#     'bcell_follicular',
#     'bcell_regulatory',
#     'tcell_general',
#     'tcell_activated',
#     'tcell_effector',
#     'tcell_regulatory',
#     'tcell_exhausted',
#     'tcell_helper_1',
#     'tcell_helper_2',
#     'tcell_naive',
#     'tcell_cytotoxic',
#     'tcell_memory',
#     'tcell_helper_17',
#     'monocyte_inflammatory',
#     'monocyte_resident',
#     'monocyte_CD14',
#     'monocyte_FCGR3A',
#     'macrophages',
#     'nk_cell',
#     'DC',
#     'pDC',
#     'megakaryocyte',
#     'erythrocyte',
#     'neutrophil',
#     'eosionophil',
#     'basophil',
#     'mast',
#     'hsc'
#     
#   )
#   feature_list <- list(
#     list('CD19','IL2RA','CD30'),
#     list('CD27','CD38','SDC1','SLAMF7','IL6'),
#     list('MS4A1','CD27','CD40','CD80','PDCD1LG2', 'CXCR3','CXCR4','CXCR5','CXCR6'),
#     list('CD1A','CR2','CD37','NOTCH2'),
#     list('CR2','CD22','FCER2'),
#     list('CD1A','CD5','CR2','CD24','TLR4','IL10','TGFB1'),
#     list('CD3D','CD3E','CD3G','PTPRC'),
#     list('CD69','IL2RA'),
#     list('CD3D','B3GAT1','PDCD1','FAS','CCR7'),
#     list('CD4','IL2RA','FOXP3','SMAD3','STAT5A','STAT5B','IL10'),
#     list('CD3D','PDCD1','FAS','CCR7','SELL','LAG3','HAVCR2','TIGIT','ENTPD1'),
#     list('CD4','CXCR3','TBX21','STAT1','STAT6','IFNG'),
#     list('CD4','CCR4','PTGDR2','GATA3','STAT5','STAT6','IL4','IL5','IL13'),
#     list('IL7R','S100A4','CD3D','SELL','CD27'),
#     list('CD8A', 'CD8B'),
#     list('IL7R','CD3D','CCR7','SELL'),
#     list('CD4','CCR6','RORC','RORA','STAT3','IL17A','IL17F'),
#     list('CCR2'),
#     list('CXCR1'),
#     list('CD14', 'FCN1', 'LYZ'),
#     list('CD14','FCGR3A','MS4A7'),
#     list('CD68','CCR5','TFRC','ITGAM','FCGR1A','CSF1R','MRC1','CD163'),
#     list('NKG7','GNLY','NCR1','KLRD1','NCAM1'),
#     list('FCER1A', 'ITGAX', 'CD83', 'THBD','CD209','CD1C', 'LYZ'),
#     list('IL3RA','CLEC4C','NRP1'),
#     list('PPBP', 'ITGA2B','GP9', 'GP1BA', 'ITGB3'),
#     list('GYPA', 'BLVRB', 'HBB', 'HBA1'),
#     list('FUT4','ITGAM','FCGR3A','ITGB2','FCGR2A','CD44','CD55'),
#     list('IL5RA','CCR3','EMR1','ADGRE1','SIGLEC8'),
#     list('IL3RA', 'ENPP3'),
#     list('KIT','CD33','TPSAB1', 'CD9'),
#     list('CD34','THY1', 'CDK6')
#     
#   )
#   
#   invisible(markers <- FindAllMarkers(data))
# 
#   for(i in seq_len(length(feature_list))){
#     cell_type = cell_list[i]
#     x <- unlist(feature_list[i])
#     gene_list = (x[x %in% markers$gene])  
#     # FeaturePlot gives error if no genes are found, so we have to check if they are included in the highly variable genes
#     if (length(gene_list > 0)){
#       print(paste0( cell_type,': Found'))
#       print(x)
#       print( FeaturePlot(sample, features = c(x)))
#       
#     }else {
#       print(paste0( cell_type,': Not Found'))
#       print(x)
#     }
#     print('')
#     
#   }
```

```{r}
cluster_IDs <- sampleParam$Cluster_IDs[sampleParam['Sample'] == sample_name]
cluster_IDs = unlist(strsplit(cluster_IDs, ", "))
  
names(cluster_IDs) <- levels(data)
data <- RenameIdents(data, cluster_IDs)
DimPlot(data, reduction = "umap", label = TRUE, pt.size = 0.5) + NoLegend()
```

# Name Clusters
```{r}
cluster_IDs <- sampleParam$Cluster_IDs[sampleParam['Sample'] == sample_name]
cluster_IDs = unlist(strsplit(cluster_IDs, ", "))

#Rename clusters
names(cluster_IDs) <- levels(data)
data <- RenameIdents(data,cluster_IDs)

#Add this to metadata
data@meta.data$renamed_clusters <- data@active.ident

DimPlot(data, reduction = "umap", label = TRUE, pt.size = 0.5) + NoLegend()
```

# Save Data
```{r}
save(data,file=paste0(folder,'data.Robj'))
```


```{r}

```